collections/_posts/2022-09-19-typelevel-native.md

```diff
@@ -0,0 +1,284 @@
+
+```c
+#include <sys/epoll.h>
+
+int epoll_create1(int flags);
+
+int epoll_ctl(int epfd, int op, int fd, struct epoll_event *event);
+
+int epoll_wait(int epfd, struct epoll_event *events,
+                      int maxevents, int timeout);
+```
+
+After creating an epoll instance (identified by a file descriptor) we can register sockets (also identified by file descriptors) with `epoll_ctl`. Typically we will register to be notified of the “read-ready” (`EPOLLIN`) and “write-ready” (`EPOLLOUT`) events on that socket. Finally, the actual polling is implemented with `epoll_wait`, which sleeps until the next I/O event is ready or the `timeout` expires. Thus we can use it to implement a  `PollingExecutorScheduler`.
+
+As previously mentioned, these sorts of polling APIs are ubiquitous and not just for working directly with sockets. For example, [libcurl](https://curl.se/libcurl/) (the C library behind the well-known CLI) exposes a function for polling for I/O on all ongoing HTTP requests.
+
+```c
+#include <curl/curl.h>
+ 
+CURLMcode curl_multi_poll(CURLM *multi_handle,
+                          struct curl_waitfd extra_fds[],
+                          unsigned int extra_nfds,
+                          int timeout_ms,
+                          int *numfds);
+```
+
+Indeed, this function underpins the `CurlExecutorScheduler` in [http4s-curl].
+
+On macOS and BSDs the [kqueue] API plays an analogous role to epoll. We will not talk about Windows today :)
+
+**Long story short, I did not want the FS2 codebase to absorb all of this cross-OS complexity.** So in collaboration with Lee Tibbert we repurposed my cheeky [epollcat] experiment into an actual library implementing JDK NIO APIs (specifically, [`AsynchronousSocketChannel`] and friends). Since these are the same APIs used by the JVM implementation of `fs2-io`, it actually enables the `Socket` code to be completely shared with Native.
+
+[epollcat] implements an `EpollExecutorScheduler` for Linux and a `KqueueExecutorScheduler` for macOS. They additionally provide an API for monitoring a socket file descriptor for read-ready and write-ready events.
+
+```scala
+def monitor(fd: Int, reads: Boolean, writes: Boolean)(
+  cb: EventNotificationCallback
+): Runnable // returns a `Runnable` to un-monitor the file descriptor
+
+trait EventNotificationCallback {
+  def notifyEvents(readReady: Boolean, writeReady: Boolean): Unit
+}
+```
+
+These are then used to implement the callback-based `read` and `write` methods of the JDK `AsynchronousSocketChannel`.
+
+It is worth pointing out that the JVM actually implements `AsynchronousSocketChannel` with an event loop as well. The difference is that on the JVM, this event loop is used only for I/O and runs on a separate thread from the compute pool used for fibers and the scheduler thread used for timers. Meanwhile, epollcat is an example of an I/O integrated runtime where fibers, timers, and I/O are all interleaved on a single thread.
```

Comment:
> epollcat is an example of an I/O integrated runtime where fibers, timers, and I/O are all interleaved on a single thread

Is it too early to comment on why this might be a better setup for cats-effect? Might it mean less thread switching?

edit: Is it perhaps worth linking to https://github.com/typelevel/cats-effect/discussions/3070 ?

Reply:
Yes, I linked to this the first time that it's mentioned. I can link again here.